man/man1/ipcs.1
.\" @(#)ipcs.1 1.1 92/07/30 SMI; from System V 6.2
.TH IPCS 1 "9 September 1987"
.SH NAME
ipcs \- report interprocess communication facilities status
.SH SYNOPSIS
.B ipcs
[
.I primitives
]
.SH DESCRIPTION
.IX  "interprocess communication"  ipcs  ""  \fLipcs\fP
.IX  "ipcs command" "" "\fLipcs\fR \(em display interprocess communication status"
.B ipcs
prints information about active interprocess communication facilities
as specified by the
.I primitives
shown below.
If no
.I primitives
are given,
information is printed in short format for message queues,
shared memory,
and semaphores that are currently active in the system.
.SS Command \- Line Primitives
.LP
If any of the
.I primitives
.BR \-m ,
.BR \-q ,
or
.B \-s
are specified,
information about only indicated facilities is printed.
If none of these are specified,
information about all three is printed.
.TP 15
.B \-q
Print information about active message queues.
.TP
.B \-m
Print information about active shared memory segments.
.TP
.B \-s
Print information about active semaphores.
.TP
.B \-b
Print the currently allowed size information.
(Maximum number of bytes in messages on queue for message queues,
size of segments for shared memory,
and number of semaphores in each set for semaphores.)
See below for the meaning of columns in a listing.
.TP
.B \-c
Print creator's login name and group name.
See below.
.TP
.B \-o
Print information on outstanding usage.
(Number of messages on queue and total number
of bytes in messages on queue for
message queues and number of processes
attached to shared memory segments.)
.TP
.B \-p
Print process number information.
(Process
.SM ID
of last process to send a message and process
.SM ID
of last process to receive a message on message queues and process
.SM ID
of creating process and process
.SM ID
of last process to attach or detach on shared memory segments)\
See below.
.TP
.B \-t
Print time information.
(Time of the last control operation that changed the access permissions for
all facilities.
Time of last
.B msgsnd
and last
.B msgrcv
(see
.BR msgop (2))
on message queues,
last
.B shmat
and last
.B shmdt
(see
.BR shmop (2))
on shared memory,
last
.BR semop (2)
on semaphores.)
See below.
.TP
.B \-a
Use all display
.IR primitive s.
(This is a shorthand notation for
.BR \-b ,
.BR \-c ,
.BR \-o ,
.BR \-p ,
and
.BR \-t .)
.TP
.BI \-C " corefile"
Use the file
.I corefile
in place of
.BR /dev/mem .
.TP
.BI \-N " namelist"
The argument will be taken as the name of an alternate
.I filenamelist
.RB ( /vmunix
is the default).
.PP
The column headings and the meaning of the columns in an
.B ipcs
listing
are given below; the letters in parentheses indicate the
.I primitives
that cause the corresponding heading to appear;
.B all
means that the heading
always appears.
Note: these
.I primitives
only determine
what information is provided for each facility; they do
.I not
determine which facilities will be listed.
.ta .75i
.ne 5
.TP 16
.SM
.BR T\*S "	(all)"
Type of the facility:
.RS 20
.PD 0
.TP 6
.B q
message queue
.TP
.B m
shared memory segment
.TP
.B s
semaphore
.PD
.RE
.TP
.SM
.BR ID\*S "	(all)"
The identifier for the facility entry.
.br
.ne 5
.TP
.SM
.BR KEY\*S "	(all)"
The key used as an argument to
.BR msgget (2),
.BR semget (2),
or
.BR shmget (2)
to create the facility entry.
(Note:
The key of a shared memory segment is changed to
.SM
.B IPC_PRIVATE
when the segment has been removed until all processes attached to the segment
detach it.)
.TP
.SM
.BR MODE\*S "	(all)"
The facility access modes and flags:
The mode consists of 11 characters that are interpreted as follows:
.ne 4
.RS
.IP
The first two characters are:
.RS
.PD 0
.TP 4
.B R
If a process is waiting on a
.BR msgrcv .
.TP
.B S
If a process is waiting on a
.BR msgsnd .
.TP
.B D
If the associated shared memory segment has been removed.
It will disappear when the last process attached to the segment
detaches it.
.TP
.B C
If the associated shared memory segment is to be cleared when the
first attach is executed.
.TP
.B \-
If the corresponding special flag is not set.
.RE
.PD
.IP
The next 9 characters are interpreted as three sets of three bits each.
The first set refers to the owner's permissions;
the next to permissions of others in the user-group of the facility entry;
and the last to all others.
Within each set, the first character indicates permission to read,
the second character indicates permission to write
or alter the facility entry,
and the last character is currently unused.
.IP
The permissions are indicated as follows:
.RS
.PD 0
.TP 4
.B r
If read permission is granted.
.TP
.B w
If write permission is granted.
.TP
.B a
If alter permission is granted.
.TP
.B \(em
If the indicated permission is
.I not
granted.
.RE
.RE
.PD
.TP
.SM
.BR OWNER\*S "	(all)"
The login name of the owner of the facility entry.
.TP
.SM
.BR GROUP\*S "	(all)"
The group name of the group of the owner of the facility entry.
.TP
.SM
.BR CREATOR\*S "	(a,c)"
The login name of the creator of the facility entry.
.TP
.SM
.BR CGROUP\*S "	(a,c)"
The group name of the group of the creator of the facility entry.
.TP
.SM
.BR CBYTES\*S "	(a,o)"
The number of bytes in messages currently outstanding on the associated
message queue.
.TP
.SM
.BR QNUM\*S "	(a,o)"
The number of messages currently outstanding on the associated message queue.
.TP
.SM
.BR QBYTES\*S "	(a,b)"
The maximum number of bytes allowed in messages outstanding on the associated
message queue.
.TP
.SM
.BR LSPID\*S "	(a,p)"
The process
.SM ID
of the last process to send a message to the associated queue.
.TP
.SM
.BR LRPID\*S "	(a,p)"
The process
.SM ID
of the last process to receive a message from the associated queue.
.TP
.SM
.BR STIME\*S "	(a,t)"
The time the last message was sent to the associated queue.
.TP
.SM
.BR RTIME\*S "	(a,t)"
The time the last message was received from the associated queue.
.TP
.SM
.BR CTIME\*S "	(a,t)"
The time when the associated entry was created or changed.
.TP
.SM
.BR NATTCH\*S "	(a,o)"
The number of processes attached to the associated shared memory segment.
.TP
.SM
.BR SEGSZ\*S "	(a,b)"
The size of the associated shared memory segment.
.TP
.SM
.BR CPID\*S "	(a,p)"
The process
.SM ID
of the creator of the shared memory entry.
.TP
.SM
.BR LPID\*S "	(a,p)"
The process
.SM ID
of the last process to attach or detach the shared memory segment.
.TP
.SM
.BR ATIME\*S "	(a,t)"
The time the last attach was completed to the associated shared memory
segment.
.br
.ne 6
.TP
.SM
.BR DTIME\*S "	(a,t)"
The time the last detach was completed on the associated shared memory
segment.
.TP
.SM
.BR NSEMS\*S "	(a,b)"
The number of semaphores in the set associated with the semaphore entry.
.TP
.SM
.BR OTIME\*S "	(a,t)"
The time the last semaphore operation was completed on the set associated
with the semaphore entry.
.DT
.PD
.SH FILES
.PD 0
.TP 20
.B /vmunix
system namelist
.TP
.B /dev/mem
memory
.TP
.B /etc/passwd
user names
.TP
.B /etc/group
group names
.PD
.SH SEE ALSO
.BR ipcrm (1),
.BR msgop (2),
.BR semctl (2),
.BR semget (2),
.BR semop (2),
.BR shmctl (2),
.BR shmget (2),
.BR shmop (2)
.SH BUGS
Things can change while
.B ipcs
is running; the picture it gives is only a close
approximation to reality.


usr.bin/ipcs.c
#ifndef lint
static	char sccsid[] = "@(#)ipcs.c 1.1 92/07/30 SMI"; /* from SYSV.2 1.5 */
#endif

/*
**	ipcs - IPC status
**	Examine and print certain things about message queues, semaphores,
**		and shared memory.
**/

#include	<sys/types.h>

#define KERNEL
#include	<sys/ipc.h>
#include	<sys/msg.h>
#include	<sys/sem.h>
#include	<sys/shm.h>
#undef KERNEL

#include	<nlist.h>
#include	<fcntl.h>
#include	<time.h>
#include	<grp.h>
#include	<pwd.h>
#include	<stdio.h>
#include	<kvm.h>

#define	TIME	0
#define	MSG	1
#define	SEM	2
#define	SHM	3
#define	MSGINFO	4
#define	SEMINFO	5
#define	SHMINFO	6

struct nlist nl[] = {		/* name list entries for IPC facilities */
	{"_time"},
	{"_msgque"},
	{"_sema"},
	{"_shmem"},
	{"_msginfo"},
	{"_seminfo"},
	{"_shminfo"},
	{NULL}
};
char	chdr[] = "T     ID     KEY        MODE       OWNER    GROUP",
				/* common header format */
	chdr2[] = "  CREATOR   CGROUP",
				/* c option header format */
	*name = NULL,		/* default name list: use live kernel */
	*mem = NULL,		/* default memory file: use live kernel */
	opts[] = "abcmopqstC:N:";/* allowable options for getopt */
extern char	*optarg;	/* arg pointer for getopt */
int		bflg,		/* biggest size:
					segsz on m; qbytes on q; nsems on s */
		cflg,		/* creator's login and group names */
		mflg,		/* shared memory status */
		oflg,		/* outstanding data:
					nattch on m; cbytes, qnum on q */
		pflg,		/* process id's: lrpid, lspid on q;
					cpid, lpid on m */
		qflg,		/* message queue status */
		sflg,		/* semaphore status */
		tflg,		/* times: atime, ctime, dtime on m;
					ctime, rtime, stime on q;
					ctime, otime on s */
		err;		/* option error count */
extern int	optind;		/* option index for getopt */
kvm_t		*kd;		/* kernel descriptor */

extern char		*ctime();
extern struct group	*getgrgid();
extern struct passwd	*getpwuid();
extern long		lseek();

main(argc, argv)
int	argc;	/* arg count */
char	**argv;	/* arg vector */
{
	register int	i,	/* loop control */
			o;	/* option flag */
	time_t		time;	/* date in memory file */
	struct shmid_ds	mds;	/* shared memory data structure */
	struct shminfo shminfo;	/* shared memory information structure */
	struct msqid_ds	qds;	/* message queue data structure */
	struct msginfo msginfo;	/* message information structure */
	struct semid_ds	sds;	/* semaphore data structure */
	struct seminfo seminfo;	/* semaphore information structure */
	long		offset;	/* used for indirect seeks */
	char		hostnm[64];	/* used for gethostname() */

	/* Go through the options and set flags. */
	while((o = getopt(argc, argv, opts)) != EOF)
		switch(o) {
		case 'a':
			bflg = cflg = oflg = pflg = tflg = 1;
			break;
		case 'b':
			bflg = 1;
			break;
		case 'c':
			cflg = 1;
			break;
		case 'C':
			mem = optarg;
			break;
		case 'm':
			mflg = 1;
			break;
		case 'N':
			name = optarg;
			break;
		case 'o':
			oflg = 1;
			break;
		case 'p':
			pflg = 1;
			break;
		case 'q':
			qflg = 1;
			break;
		case 's':
			sflg = 1;
			break;
		case 't':
			tflg = 1;
			break;
		case '?':
			err++;
			break;
		}
	if(err || (optind < argc)) {
		fprintf(stderr,
		    "usage:  ipcs [-abcmopqst] [-C corefile] [-N namelist]\n");
		exit(1);
	}
	if((mflg + qflg + sflg) == 0)
		mflg = qflg = sflg = 1;

	/* initialize for Kernel VM access. */
	if((kd = kvm_open(name, mem, NULL, O_RDONLY, "ipcs")) == NULL) {
		exit(1);
	}
	if (kvm_nlist(kd, nl) == -1) {
		fprintf(stderr, "ipcs:  bad namelist\n");
		exit(1);
	}

	if (mem == NULL) {
		gethostname(hostnm, sizeof(hostnm));
		mem = hostnm;
	}
	readsym(TIME, &time, sizeof(time));
	printf("IPC status from %s as of %s", mem, ctime(&time));

	/* Print Message Queue status report. */
	if(qflg) {
		i = 0;
		if ((nl[MSGINFO].n_type != 0) &&
		  (readsym(MSGINFO, &msginfo, sizeof(msginfo)),
		  (msginfo.msgmni != 0))) {
		    /* find the real address of the id pool */
		    readsym(MSG, &offset, sizeof(offset));

		    printf("%s%s%s%s%s%s\nMessage Queues:\n", chdr,
			    cflg ? chdr2 : "",
			    oflg ? " CBYTES  QNUM" : "",
			    bflg ? " QBYTES" : "",
			    pflg ? " LSPID LRPID" : "",
			    tflg ? "   STIME    RTIME    CTIME " : "");
		    while(i < msginfo.msgmni) {
			reade(offset, &qds, sizeof(qds));
			offset += sizeof(qds);
			if(!(qds.msg_perm.mode & IPC_ALLOC)) {
				i++;
				continue;
			}
			hp('q', "SRrw-rw-rw-", &qds.msg_perm, i++, msginfo.msgmni);
			if(oflg)
				printf("%7u%6u", qds.msg_cbytes, qds.msg_qnum);
			if(bflg)
				printf("%7u", qds.msg_qbytes);
			if(pflg)
				printf("%6u%6u", qds.msg_lspid, qds.msg_lrpid);
			if(tflg) {
				tp(qds.msg_stime);
				tp(qds.msg_rtime);
				tp(qds.msg_ctime);
			}
			printf("\n");
		    }
		} else {
		    printf("Message Queue facility not in system.\n");
		}
	}

	/* Print Shared Memory status report. */
	if(mflg) {
		i = 0;
		if ((nl[SHMINFO].n_type != 0) &&
		  (readsym(SHMINFO, &shminfo, sizeof(shminfo)),
		  (shminfo.shmmni != 0))) {
		    /* find the real address of the id pool */
		    readsym(SHM, &offset, sizeof(long));

		    if(oflg || bflg || tflg || !qflg || !nl[MSG].n_value)
			    printf("%s%s%s%s%s%s\n", chdr,
				    cflg ? chdr2 : "",
				    oflg ? " NATTCH" : "",
				    bflg ? "  SEGSZ" : "",
				    pflg ? "  CPID  LPID" : "",
				    tflg ? "   ATIME    DTIME    CTIME " : "");
		    printf("Shared Memory:\n");
		    while(i < shminfo.shmmni) {
			reade(offset, &mds, sizeof(mds));
			offset += sizeof(mds);
			if(!(mds.shm_perm.mode & IPC_ALLOC)) {
				i++;
				continue;
			}
			/* IPC_STAT needed to update shm_nattch */
			if (shmctl(i + shminfo.shmmni * mds.shm_perm.seq, 
							IPC_STAT, &mds) < 0) {
				perror("shmctl");
				exit(1);
			}
			hp('m', "DCrw-rw-rw-", &mds.shm_perm, i++, shminfo.shmmni);
			if(oflg)
				printf("%7u", mds.shm_nattch);
			if(bflg)
				printf("%7d", mds.shm_segsz);
			if(pflg)
				printf("%6u%6u", mds.shm_cpid, mds.shm_lpid);
			if(tflg) {
				tp(mds.shm_atime);
				tp(mds.shm_dtime);
				tp(mds.shm_ctime);
			}
			printf("\n");
		    }
		} else {
		    printf("Shared Memory facility not in system.\n");
		}
	}

	/* Print Semaphore facility status. */
	if(sflg) {
		i = 0;
		if ((nl[SEMINFO].n_type != 0) &&
		  (readsym(SEMINFO, &seminfo, sizeof(seminfo)),
		  (seminfo.semmni != 0))) {
		    /* find the real address of the id pool */
		    readsym(SEM, &offset, sizeof(long));

		    if(bflg || tflg || (!qflg || !nl[MSG].n_value) &&
			(!mflg || !nl[SHM].n_value))
			    printf("%s%s%s%s\n", chdr,
				    cflg ? chdr2 : "",
				    bflg ? " NSEMS" : "",
				    tflg ? "   OTIME    CTIME " : "");
		    printf("Semaphores:\n");
		    while(i < seminfo.semmni) {
			reade(offset, &sds, sizeof(sds));
			offset += sizeof(sds);
			if(!(sds.sem_perm.mode & IPC_ALLOC)) {
				i++;
				continue;
			}
			hp('s', "--ra-ra-ra-", &sds.sem_perm, i++, seminfo.semmni);
			if(bflg)
				printf("%6u", sds.sem_nsems);
			if(tflg) {
				tp(sds.sem_otime);
				tp(sds.sem_ctime);
			}
			printf("\n");
		    }
		} else {
		    printf("Semaphore facility not in system.\n");
		}
	}
	exit(0);
	/* NOTREACHED */
}

/*
**	readsym - read kernel symbol value with error exit
*/

readsym(n, b, s)
int	n,	/* nl offset */
	s;	/* size */
char	*b;	/* buffer address */
{
	if (nl[n].n_type == 0) {
		fprintf(stderr, "ipcs:  '%s' not in namelist\n", nl[n].n_name);
		exit(1);
	}
	return (reade(nl[n].n_value, b, s));
}

/*
**	reade - read with error exit
*/

reade(a, b, s)
u_long	a;	/* address */
int	s;	/* size */
char	*b;	/* buffer address */
{
	if (kvm_read(kd, a, b, s) != s) {
		fprintf(stderr, "ipcs:  kernel read error\n");
		exit(1);
	}
	return (0);
}

/*
**	hp - common header print
*/

hp(type, modesp, permp, slot, slots)
char				type,	/* facility type */
				*modesp;/* ptr to mode replacement characters */
register struct ipc_perm	*permp;	/* ptr to permission structure */
int				slot,	/* facility slot number */
				slots;	/* # of facility slots */
{
	register int		i,	/* loop control */
				j;	/* loop control */
	register struct group	*g;	/* ptr to group group entry */
	register struct passwd	*u;	/* ptr to user passwd entry */

	printf("%c%7d%s%#8.8x ", type, slot + slots * permp->seq,
		permp->key ? " " : " 0x", permp->key);
	for(i = 02000;i;modesp++, i >>= 1)
		printf("%c", (permp->mode & i) ? *modesp : '-');
	if((u = getpwuid(permp->uid)) == NULL)
		printf("%9d", permp->uid);
	else
		printf("%9.8s", u->pw_name);
	if((g = getgrgid(permp->gid)) == NULL)
		printf("%9d", permp->gid);
	else
		printf("%9.8s", g->gr_name);
	if(cflg) {
		if((u = getpwuid(permp->cuid)) == NULL)
			printf("%9d", permp->cuid);
		else
			printf("%9.8s", u->pw_name);
		if((g = getgrgid(permp->cgid)) == NULL)
			printf("%9d", permp->cgid);
		else
			printf("%9.8s", g->gr_name);
	}
}

/*
**	tp - time entry printer
*/

tp(time)
time_t	time;	/* time to be displayed */
{
	register struct tm	*t;	/* ptr to converted time */

	if(time) {
		t = localtime(&time);
		printf(" %2d:%2.2d:%2.2d", t->tm_hour, t->tm_min, t->tm_sec);
	} else
		printf(" no-entry");
}
